### PR TITLE
Add support for nodeSelector on deployments

### PR DESCRIPTION
**Pull request by kody-abe**
This adds support for nodeSelector on deployments.

I am pretty new to operators and ansible templates, so I think this will work. Certainly, will need someone more experienced to validate first.

I am trying to get the docker to build locally and testin my cluster.

### CODE REVIEW

**Comment by kody-abe** [comment]
I have tested this and it works as expected! 🎉

**Comment by slopezz** [comment]
Thanks for contributing @kody-abe , let me check and test it before merging it!

File: roles/prometheusexporter/templates/deployment.yml.j2 [hunk 1]
```diff
@@ -130,3 +130,6 @@ spec:
           - key: config.yml
             path: config.yml
 {% endif %}
+{% if node_selector is defined %}
+      nodeSelector: {{ node_selector }}
+{% endif %}
```
Comment (slopezz) [inline roles/prometheusexporter/templates/deployment.yml.j2:135]: I have tested it and it does not work, running the operator locally with a modified CR from examples directory (and adding the new `nodeSelector` field), it fails with:
```json
 TASK [Manage Deployment for PrometheusExporter staging-backend-redis on Namespace prom-exporter] ******************************** 
fatal: [localhost]: FAILED! => {"changed": false, "error": 422, "msg": "Failed to patch object: {\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"Deployment.apps \\\"prometheus-exporter-redis-staging-backend-redis\\\" is invalid: [spec.template.spec.nodeSelector: Invalid value: \\\"u'node'\\\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]'), spec.template.spec.nodeSelector: Invalid value: \\\"u'test'\\\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')]\",\"reason\":\"Invalid\",\"details\":{\"name\":\"prometheus-exporter-redis-staging-backend-redis\",\"group\":\"apps\",\"kind\":\"Deployment\",\"causes\":[{\"reason\":\"FieldValueInvalid\",\"message\":\"Invalid value: \\\"u'node'\\\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')\",\"field\":\"spec.template.spec.nodeSelector\"},{\"reason\":\"FieldValueInvalid\",\"message\":\"Invalid value: \\\"u'test'\\\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')\",\"field\":\"spec.template.spec.nodeSelector\"}]},\"code\":422}\n", "reason": "Unprocessable Entity", "status": 422}
```

After testing a way to do so, this is the code that works:
```suggestion
{% if node_selector is defined %}
      nodeSelector:
{% for key, value in node_selector.items() %}
        {{ key | safe }}: {{ value | safe }}
{% endfor %}
{% endif %}
```

**Review by slopezz — CHANGES_REQUESTED** [review]
Thanks for the contribution @kody-abe 

I have made a few suggestions, can you please check them?

There is just a drawback, because of ansible operator converts **hyphens** to **underscores**, if the key-value pairs contain a **hypen** on the nodeSelector `key` at the Custom Resource, example:
```yaml
apiVersion: monitoring.3scale.net/v1alpha1
kind: PrometheusExporter
metadata:
  name: staging-backend-redis
spec:
...
  nodeSelector:
    node: test
    key-hyphen: value-hypen   ## That one, key-hyphen
    key_under_score: value_under_score
```

The operator will convert `keys` with hyphen to `keys` with underscore (`values` can contain a hyphen with no problem), this is what the final deployment will look like:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: prometheus-exporter-redis-staging-backend-redis
spec:
  template:
    spec:
...
      nodeSelector:
        key_hyphen: value-hypen  ## That one, key has been modified from key-hyphen to key_hyphen
        key_under_score: value_under_score
        node: test
```

Knowing that simple limitation, I think it is fair enough going that way.

In addition, you need to modify the CRD at `deploy/crds/monitoring.3scale.net_prometheusexporters_crd.yaml`, with the new `nodeSelector` map (description will concide with the one at the docs):

```yaml
            nodeSelector:
              additionalProperties:
                type: string
              description: Map of nodeSelector key-value pairs
              type: object
```

**Comment by kody-abe** [comment]
@slopezz Well, the lack of hypen support here is interesting. I tried a few things and found now way around ansible operator outputting it like that.I have taken care of the rest of the comments.

**Review by slopezz — APPROVED** [review]
Thanks @kody-abe , I have tested and everything works as expected.

I will open a new PR with a new operator release `0.2.4` with updated operator CSVs.. so I can publish it into [OperatorHub.io](https://operatorhub.io/operator/prometheus-exporter-operator)

**Pull request merged**